XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,419 @@
+Node's implementation details, and is in fact intentional. Observing how
+operations are performed, not simply how they appear to perform from the public
+API, is a key point in its usability. For a real world example, a user was
+having difficulty figuring out why their `http.get()` calls were taking so
+long. Because `AsyncHooks` exposed the `dns.lookup()` request to resolve the
+host being made by the `net` module it was trivial to write a performance
+measurement that exposed the hidden culprit. Which was that DNS resolution was
+taking much longer than expected.
+
+A small amount of abstraction is necessary to accommodate the conceptual nature
+of the API. For example, the `HTTPParser` is not destructed but instead placed
+into an unused pool to be used later. Even so, at the time it is placed into
+the pool  the `destroy()` callback will run and then the id assigned to that
+instance will be removed. If that resource is again requested then it will be
+assigned a new id and will run `init()` again.
+
+
+## Goals
+
+The intent for the initial release is to provide the most minimal set of API
+hooks that don't inhibit module authors from writing tools addressing anything
+in this problem space. In order to remain minimal, all potential features for
+initial release will first be judged on whether they can be achieved by the
+existing public API. If so then such a feature won't be included.
+
+If the feature cannot be done with the existing public API then discussion will
+be had on whether requested functionality should be included in the initial
+API. If so then the best course of action to include the capabilities of the
+request will be discussed and performed. Meaning, the resulting API may not be
+the same as initially requested, but the user will be able to achieve the same
+end result.
+
+Performance impact of `AsyncWrap` should be zero if not being used, and near
+zero while being used. The performance overhead of `AsyncHook` callbacks
+supplied by the user should account for essentially all the performance
+overhead.
+
+
+## Terminology
+
+Because of the potential ambiguity for those not familiar with the terms "handle"
+and "request" they should be well defined.
+
+Handles are a reference to a system resource. The resource handle can be a
+simple identifier, such as a file descriptor, or it can be a pointer that
+allows access to further information, such as `uv_tcp_t`. Realize that the
+latter still relies on a file descriptor, but has been encapsulated in a data
+structure containing additional information that allows the application to
+work with it.
+
+Requests are short lived data structures created to accomplish one task. The
+callback for a request should always and only ever fire one time. When the
+assigned task has either completed or encountered an error. Requests are used
+by handles to perform work. Such as accepting a new connection or writing data
+to disk.
+
+
+## API
+
+### Overview
```

Comment:
Could you put a note in here somehow that mentions that further details are provided in the section below? Because you go into a fair bit of detail in the overview it feels like you're giving a complete picture (hence some of the questions you've been asked in this PR). Something like "Further details on each element of the API can be found in sections below".